File: app/src/main/java/configgen/data/DataUtil.java
package configgen.data;

import configgen.util.FileNameUtil;

import java.nio.file.Path;
import java.util.ArrayList;
import java.util.List;

import static configgen.data.DataUtil.FileFmt.*;

public class DataUtil {

    public enum FileFmt {
        TXT_AS_TSV,
        CSV,
        EXCEL,
        CFG,
        JSON,
    }


    public static boolean isFileIgnored(Path path) {
        return path.toFile().isHidden() || path.getFileName().toString().startsWith("~");
    }

    public static FileFmt getFileFormat(Path path) {
        String fileName = path.getFileName().toString();
        String ext = "";
        int i = fileName.lastIndexOf('.');
        if (i >= 0) {
            ext = fileName.substring(i + 1).toLowerCase();
        }

        return switch (ext) {
            case "txt" -> TXT_AS_TSV; // 之前的tsv文件存成了txt
            case "csv" -> CSV;
            case "xls", "xlsx" -> EXCEL;
            case "cfg" -> CFG;
            case "json" -> JSON;
            default -> null;
        };
    }

    public record TableNameIndex(String tableName,
                                 int index) {
    }

    public static TableNameIndex getTableNameIndex(Path filePath, String sheetName) {
        Path path;
        if (filePath.getParent() != null) {
            path = filePath.getParent().resolve(sheetName);
        } else {
            path = Path.of(sheetName);
        }
        return getTableNameIndex(path);
    }

    //现在都小写了，要是讲究的话，应该是路径小写，sheetName不改
    public static TableNameIndex getTableNameIndex(Path filePath) {
        List<String> codeNames = new ArrayList<>();
        for (Path path : filePath) {
            String fileName = path.getFileName().toString();
            String codeName = FileNameUtil.getCodeName(fileName);
            if (codeName == null) {
                return null;
            }
            codeNames.add(codeName);
        }
        String fullName = String.join(".", codeNames);

        String tableName;
        int index;
        int i = fullName.lastIndexOf("_");
        if (i < 0) {
            tableName = fullName.trim();
            index = 0;
        } else {
            String postfix = fullName.substring(i + 1).trim();
            try {
                index = Integer.parseInt(postfix);
                tableName = fullName.substring(0, i).trim();
            } catch (NumberFormatException ignore) {
                tableName = fullName.trim();
                index = 0;
            }
        }
        return new TableNameIndex(tableName, index);
    }



    public static String getJsonTableDirName(String tableName) {
        return "_" + tableName.replace(".", "_");
    }

    /// 从 "_" 开头的 JSON 目录名提取子表名。
    /// "_buff" → "buff"；非 "_" 开头、_后不是英文字母、或含中文返回 null。
    private static String subTableNameIfJsonDir(String dirName) {
        if (!dirName.startsWith("_")) {
            return null;
        }
        String sub = dirName.substring(1);
        // _后要是英文字母
        if (sub.isEmpty() || FileNameUtil.isFirstNotAzChar(sub)) {
            return null;
        }

        // 不能含中文
        int hanIdx = FileNameUtil.findFirstHanIndex(sub);
        if (hanIdx != -1) {
            return null;
        }
        return sub;
    }

    public static String getTableNameIfTableDirForJson(String dirName) {
        String sub = subTableNameIfJsonDir(dirName);
        if (sub == null) {
            return null;
        }
        return sub.replace("_", ".");
    }

    /// 从嵌套的 JSON 子目录名提取表名的后半部分。
    /// 如 "_buff" → "buff"，"_instancelogic" → "instancelogic"。
    /// 非 "_" 开头或不符合规则返回 null。
    public static String getSubTableNameIfJsonSubDir(String subDirName) {
        return subTableNameIfJsonDir(subDirName);
    }

    public static boolean isTableDirForJson(String dirName) {
        return subTableNameIfJsonDir(dirName) != null;
    }

}

File: packages/data/src/DataUtil.ts
/**
 * DataUtil — TypeScript port of Java `configgen.data.DataUtil`.
 *
 * File format detection, table name extraction, and JSON directory helpers.
 */

import * as path from 'path';
import { getCodeName } from '@cfggen/shared';

// ---------------------------------------------------------------------------
// FileFmt
// ---------------------------------------------------------------------------

export enum FileFmt {
  TXT_AS_TSV,
  CSV,
  EXCEL,
  CFG,
  JSON,
}

export function getFileFormat(filePath: string): FileFmt | null {
  const fileName = path.basename(filePath);
  const ext = path.extname(fileName).slice(1).toLowerCase();
  switch (ext) {
    case 'txt': return FileFmt.TXT_AS_TSV;
    case 'csv': return FileFmt.CSV;
    case 'xls':
    case 'xlsx': return FileFmt.EXCEL;
    case 'cfg': return FileFmt.CFG;
    case 'json': return FileFmt.JSON;
    default: return null;
  }
}

export function isFileIgnored(filePath: string): boolean {
  const fileName = path.basename(filePath);
  return fileName.startsWith('~') || fileName.startsWith('.');
}

// ---------------------------------------------------------------------------
// TableNameIndex
// ---------------------------------------------------------------------------

export class TableNameIndex {
  readonly tableName: string;
  readonly index: number;

  constructor(tableName: string, index: number) {
    this.tableName = tableName;
    this.index = index;
  }
}

/**
 * Extract table name and index from a relative file path (for CSV)
 * or from a file path + sheet name (for Excel).
 *
 * The path components are converted to code names via FileNameUtil.getCodeName.
 * A trailing "_N" suffix is parsed as an index for multi-file/multi-sheet tables.
 *
 * Returns null if any path component has an invalid code name.
 */
export function getTableNameIndex(filePath: string): TableNameIndex | null;
export function getTableNameIndex(filePath: string, sheetName: string): TableNameIndex | null;
export function getTableNameIndex(filePath: string, sheetName?: string): TableNameIndex | null {
  // If sheetName provided, resolve relative to filePath's parent
  let targetPath: string;
  if (sheetName !== undefined) {
    const dir = path.dirname(filePath);
    targetPath = path.join(dir, sheetName);
  } else {
    targetPath = filePath;
  }

  // Split path into components and convert each to code name
  const parts = targetPath.split(/[\\/]/).filter((p) => p.length > 0);
  const codeNames: string[] = [];
  for (const part of parts) {
    const codeName = getCodeName(part);
    if (codeName === null) {
      return null;
    }
    codeNames.push(codeName);
  }

  const fullName = codeNames.join('.');

  // Parse trailing _N as index
  const i = fullName.lastIndexOf('_');
  if (i < 0) {
    return new TableNameIndex(fullName.trim(), 0);
  } else {
    const postfix = fullName.slice(i + 1).trim();
    const parsed = parseInt(postfix, 10);
    if (!isNaN(parsed)) {
      return new TableNameIndex(fullName.slice(0, i).trim(), parsed);
    } else {
      return new TableNameIndex(fullName.trim(), 0);
    }
  }
}

// ---------------------------------------------------------------------------
// JSON directory helpers
// ---------------------------------------------------------------------------

export function getJsonTableDirName(tableName: string): string {
  return '_' + tableName.replace(/\./g, '_');
}

export function getTableNameIfTableDirForJson(dirName: string): string | null {
  const sub = subTableNameIfJsonDir(dirName);
  if (sub === null) return null;
  return sub.replace(/_/g, '.');
}

export function getSubTableNameIfJsonSubDir(subDirName: string): string | null {
  return subTableNameIfJsonDir(subDirName);
}

export function isTableDirForJson(dirName: string): boolean {
  return subTableNameIfJsonDir(dirName) !== null;
}

function subTableNameIfJsonDir(dirName: string): string | null {
  if (!dirName.startsWith('_')) {
    return null;
  }
  const sub = dirName.slice(1);
  if (sub.length === 0 || isFirstNotAzChar(sub)) {
    return null;
  }
  // Cannot contain Chinese characters
  const hanIdx = findFirstHanIndex(sub);
  if (hanIdx !== -1) {
    return null;
  }
  return sub;
}

// These are re-exported from shared, but DataUtil in Java had them inline.
// Importing directly from shared would create a circular dependency,
// so we re-implement the minimal check here.
import { isFirstNotAzChar, findFirstHanIndex } from '@cfggen/shared';
